paimon-core/src/main/java/org/apache/paimon/utils/TagManager.java

```diff
@@ -355,9 +355,16 @@ public List<String> allTagNames() {
     }
 
     private int findIndex(Snapshot taggedSnapshot, List<Snapshot> taggedSnapshots) {
-        for (int i = 0; i < taggedSnapshots.size(); i++) {
-            if (taggedSnapshot.id() == taggedSnapshots.get(i).id()) {
-                return i;
+        int left = 0;
+        int right = taggedSnapshots.size() - 1;
+        while (left <= right) {
+            int mid = left + (right - left) / 2;
+            if (taggedSnapshots.get(mid).id() == taggedSnapshot.id()) {
+                return mid;
+            } else if (taggedSnapshots.get(mid).id() < taggedSnapshot.id()) {
+                left = mid + 1;
+            } else {
+                right = mid - 1;
```

Comment:
How about encapsulating the binary search into a general static method?

Reply:
good idea. should we add a new util class `ListUtils` or do you have some advice?